New Job resource

soltysh opened the pull request:
I'm adding new job resource as defined in #11746. I picked up the initial changes from #7380, but per @erictune suggestion I'm placing the new object in `expapi`.

@smarterclayton @derekwaynecarr @sdminonne @akaram @bprashanth fyi

soltysh committed:
New Job resource

pmorie commented:
@k8s-bot ok to test

On Wednesday, August 19, 2015, Kubernetes Bot <notifications@github.com>
wrote:

> Can one of the admins verify that this patch is reasonable to test? (reply
> "ok to test", or if you trust the user, reply "add to whitelist")
>
> If this message is too spammy, please complain to ixdy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12910#issuecomment-132557344>
> .
>

soltysh commented:
I think I've reached a state where I can focus entirely on the controller logic, and that's the plan for the next few days.

jlowdermilk commented:
I am not familiar enough with these components to be a useful reviewer. Reassigning to @davidopp to review/redirect.

davidopp commented:
Assigning to @erictune who has been shepherding the job controller work.

soltysh commented:
@erictune after talking to @smarterclayton I've decided to split the work into two separate PRs to simplify the reviews. This one contains the new job resource, validation, generated stuff and storage (1st commit). The 2nd commit contains the client API for jobs. The controller part will be in a separate PR. Does it make sense or should I split it more?

bgrant0607 commented on examples/job/job.yaml:
```diff
@@ -0,0 +1,19 @@
+kind: Job
+apiVersion: v1
+metadata:
```
Belongs in docs/user-guide rather than examples, but needs documentation

soltysh replied:
I'll leave it out for a separate PR then.

bgrant0607 commented on pkg/expapi/types.go:
```diff
@@ -264,3 +264,98 @@ type DeploymentStatus struct {
+type JobCondition struct {
+	Type               JobConditionType    `json:"type"`
+	Status             api.ConditionStatus `json:"status"`
+	LastHeartbeatTime  util.Time           `json:"lastHeartbeatTime,omitempty"`
```
Please see: https://github.com/kubernetes/kubernetes/pull/12894#issuecomment-134689111

soltysh replied:
Changed.

soltysh replied:
Does this requires changing the [docs/proposal/job.md](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/job.md) or this can be left after this reaches some reasonable final state?

bgrant0607 commented on pkg/expapi/types.go:
```diff
@@ -264,3 +264,98 @@ type DeploymentStatus struct {
+	// run at any given time. The actual number of pods running in steady state will
+	// be less than this number when ((.spec.completions - .status.successful) < .spec.parallelism),
+	// i.e. when the work left to do is less than max parallelism.
+	Parallelism *int `json:"parallelism,omitempty" description:"maximum number of pods the job should run"`
```
FYI, we're going to switch expapi to use comments rather than description tags for documentation at some point, hopefully soon. If the presubmit checks don't check expapi yet, you could make this change now.

In any case, we don't put the description tags in the internal rep (i.e., this file).

soltysh replied:
From my own tests it looks like it's not verified, so per your suggestion I'm removing those from both places. And thank you for reading those from the comment, this will save quite a lot of duplication.

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -590,6 +592,40 @@ func printReplicationControllerList(list *api.ReplicationControllerList, w io.Wr
+	if len(containers) > 0 {
+		firstContainer, containers = containers[0], containers[1:]
+	}
+	_, err := fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%d\n",
```
example output?

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -590,6 +592,40 @@ func printReplicationControllerList(list *api.ReplicationControllerList, w io.Wr
+	}
+	// Lay out all the other containers on separate lines.
+	for _, container := range containers {
+		_, err := fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\n", "", container.Name, container.Image, "", "")
```
Please read:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md

No multi-line get output. Could do that in describe, which you should implement.

I'd push the kubectl changes to another PR. This one is pretty big.

soltysh replied:
Makes sense, I'll create a separate PR for that part.

soltysh commented:
Per @bgrant0607 suggestion I've dropped kubectl related stuff (will post a separate PR for that) and example job yaml file (again separate doc PR will be posted sometime towards the end of the work).
Furthermore I've removed the description form types definitions and renamed the JobCondition's LastHeartbeatTime to LastProbeTime. Appreciate another round of review.

soltysh commented:
I realized that the client code should also land in the other PR. Fixed that as well. This should minimize the amount of code in this PR, at least a bit.

soltysh commented:
@erictune @bgrant0607 ptal

erictune commented:
What does RestartPolicyNever on a Pod template mean for the job?  Here are all the resonable options I can think of:

1. if a Container run attempt of a Pod of this Job exits with failure, stop all the Pods of the Job and don't make new ones?
2. if a Container run attempt of a Pod of this Job exits with failure, do not start any new Pods, but let the existing ones keep running.
3. if a Container run attempt of a Pod of this Job exits with failure, don't make any new Pods corresponding to "this completion", but do retry ones corresponding to other "completions".  _(I think in your initial implementation, there is no notion of which completion a Pod corresponds to, but this is something I was going to suggest for a follow on, so it is worth thinking about)_
4. if a Container run attempt of a Pod of this Job exits with failure, do keep making new Pods.

I haven't reviewed  #13259 yet, so I don't know which option you picked.  But, actually, I think none of them are satisfactory. 

The problem is that RestartPolicyNever could mean two things.

It could mean _once this container has failed, a new Pod must be created_.  Reasons for needing a new Pod include:
 - wants to run again with a clean emptyDir.
 - the other containers in the pod have a certain startup sequence which needs to be redone.
 - a volume such as Secrets or Config needs to be recreated to get a newer version of data
 - there was something wrong with one of my volumes, such as a PD, and so I am restarting to get a new one.

On the other hand, it could mean _once this container has failed, a new Pod should not be created_.  Reasons for this include:
 - The container caused some globally-visible side effect and then hit an error.  A human must undo the side effect before the next invokation of the same container, or bad things will happen.
 - The container detected an error condition which it knows that restarting will not fix.  Therefore, it is pointless to run more pods.  So don't start any to avoid waste.

erictune commented:
If we think that RestartPolicyNever means "okay to make a new Pod, but don't restart this container", then we should go with option 4 above.    However, I do think this is counterintuitive for many users, so at the very least, we need to document it well.

If we think that RestartPolicyNever means "not okay to make new Pods, because bad stuff will happen", then I guess we would need to go with options 1, 2, or 3.  However, if the job has parallelism greater than 1, there is no way we can guarantee that no new pods are started (given that notice of the failure can be delayed).

If we think that RestartPolicyNever means "prefer not to make new Pods, because they probably won't succeed", the we could go with options 1, 2, or 3. 

My current thinking is that we should go with option 4, and

erictune commented:
The problem with 4 then is do we count a completion?  If we count a completion, then the user doesn't get the restart they wanted.  But if we do not count a completion, then we won't ever terminate the job.

soltysh commented:
@erictune #13259 needs more care before you'll find an answer to your question, I'm still working on it. The general idea I had in my head when writing the proposal was 4, iow. if a Pod failed, increase JobStatus.Unsuccessful and continue creating other Pods. In the end the sum of JobStatus.Unsuccessful and JobStatus.Successful must match JobSpec.Completions. In contrast RestartPolicyOnFailure will always have that value of JobStatus.Unsuccessful equal to 0. This is mentioned in the description of that field, iirc @bgrant0607 insisted on having it explained there. Does this addresses your doubts or shall I expand either the proposal or definition on the fields?

soltysh commented:
Rebased the PR.

soltysh commented:
Rebased and addressed comments from @erictune and @bgrant0607 in #13254.

erictune commented:
I'm on vacation this week (:house: :mouse:) and will take a look next week.

soltysh commented:
Rebased against latest master to pick up the client changes, esp. the experimental part.

erictune commented on pkg/expapi/types.go:
```diff
@@ -362,3 +362,105 @@ type ThirdPartyResourceDataList struct {
+
+// Job represents the configuration of a single job.
+type Job struct {
+	api.TypeMeta   `json:",inline"`
```
Insert:
```
        // Standard object's metadata.
        // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
```

erictune commented on pkg/expapi/types.go:
```diff
@@ -362,3 +362,105 @@ type ThirdPartyResourceDataList struct {
+	api.TypeMeta   `json:",inline"`
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is a structure defining the expected behavior of a job.
```
```
        // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
```

erictune commented on pkg/expapi/types.go:
```diff
@@ -362,3 +362,105 @@ type ThirdPartyResourceDataList struct {
+	// Spec is a structure defining the expected behavior of a job.
+	Spec JobSpec `json:"spec,omitempty"`
+
+	// Status is a structure describing current status of a job.
```
ditto

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+	v1.TypeMeta   `json:",inline"`
+	v1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is a structure defining the expected behavior of a job.
```
```
        // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
```

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+	// Spec is a structure defining the expected behavior of a job.
+	Spec JobSpec `json:"spec,omitempty"`
+
+	// Status is a structure describing current status of a job.
```
```
        // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
```

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+
+// Job represents the configuration of a single job.
+type Job struct {
+	v1.TypeMeta   `json:",inline"`
```
Insert:
```
        // Standard object's metadata.
        // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
```

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+
+	// Completions specifies the desired number of successfully finished pods the
+	// job should be run with. Defaults to 1.
+	Completions *int `json:"completions,omitempty" descrption:"desired number of pods"`
```
`descrption` is misspelled, and I believe should not be used anymore.

erictune commented on pkg/expapi/types.go:
```diff
@@ -362,3 +362,105 @@ type ThirdPartyResourceDataList struct {
+
+	// Completions specifies the desired number of successfully finished pods the
+	// job should be run with. Defaults to 1.
+	Completions *int `json:"completions,omitempty" descrption:"desired number of pods"`
```
remove "descrption"

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+type JobStatus struct {
+
+	// Conditions represent the latest available observations of an object's current state.
+	Conditions []JobCondition `json:"conditions,omitempty"`
```
This should probably be `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+	// Conditions represent the latest available observations of an object's current state.
+	Conditions []JobCondition `json:"conditions,omitempty"`
+
+	// StartTime represents time when the job was started. It is not guaranteed to
```
s/started/acknowledged by the Job Manager/

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+
+	// StartTime represents time when the job was started. It is not guaranteed to
+	// be set in happens-before order across separate operations.
+	// Clients may not set this value. It is represented in RFC3339 form and is in UTC.
```
No need to state that Clients may not set this value.  In general, clients may not set JobStatus on a POST/PUT to `/expapi/v1/jobs/myjob`.   To set this, one must POST to  `/expapi/v1/jobs/myjob/status`.  I haven't looked at that part of the implementation PR, but if you know it doesn't do that validation, them please add it.  If you were copying ReplicationController, then I think you'll have got it right.

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+	// be set in happens-before order across separate operations.
+	// Clients may not set this value. It is represented in RFC3339 form and is in UTC.
+	//
+	// Populated by the system.
```
Remove remark about Populated by the system.  (I might have said to add that before... if so sorry I was wrong.)

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+
+	// CompletionTime represents time when the job was completed. It is not guaranteed to
+	// be set in happens-before order across separate operations.
+	// Clients may not set this value. It is represented in RFC3339 form and is in UTC.
```
See above

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+	// be set in happens-before order across separate operations.
+	// Clients may not set this value. It is represented in RFC3339 form and is in UTC.
+	//
+	// Populated by the system.
```
see above.

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+	//
+	// Populated by the system.
+	// Read-only.
+	StartTime util.Time `json:"startTime,omitempty"`
```
Pointer (I think Clayton already pointed this out).

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+
+// JobStatusResult represents the current status of a job.
+type JobStatusResult struct {
+	v1.TypeMeta   `json:",inline"`
```
```
  // Standard object's metadata. 
  // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
```

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+)
+
+// JobCondition describes current state of a job.
+type JobCondition struct {
```
```// Type of job condition, currently only Complete.```

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+
+// JobCondition describes current state of a job.
+type JobCondition struct {
+	Type               JobConditionType   `json:"type"`
```
```
 // Status of the condition, one of True, False, Unknown.
```

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+// JobCondition describes current state of a job.
+type JobCondition struct {
+	Type               JobConditionType   `json:"type"`
+	Status             v1.ConditionStatus `json:"status"`
```
``` // Last time the condition was checked.```

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+type JobCondition struct {
+	Type               JobConditionType   `json:"type"`
+	Status             v1.ConditionStatus `json:"status"`
+	LastProbeTime      util.Time          `json:"lastProbeTime,omitempty"`
```
```
 // Last time the condition transit from one status to another.
```

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+	Type               JobConditionType   `json:"type"`
+	Status             v1.ConditionStatus `json:"status"`
+	LastProbeTime      util.Time          `json:"lastProbeTime,omitempty"`
+	LastTransitionTime util.Time          `json:"lastTransitionTime,omitempty"`
```
```
 // (brief) reason for the condition's last transition.
```

erictune commented on pkg/expapi/v1/types.go:
```diff
@@ -350,3 +350,105 @@ type ThirdPartyResourceDataList struct {
+	Status             v1.ConditionStatus `json:"status"`
+	LastProbeTime      util.Time          `json:"lastProbeTime,omitempty"`
+	LastTransitionTime util.Time          `json:"lastTransitionTime,omitempty"`
+	Reason             string             `json:"reason,omitempty"`
```
```
  // Human readable message indicating details about last transition.
```

erictune commented on pkg/registry/job/rest.go:
```diff
@@ -0,0 +1,106 @@
+// PrepareForCreate clears the status of a job before creation.
+func (jobStrategy) PrepareForCreate(obj runtime.Object) {
+	job := obj.(*expapi.Job)
+	job.Status = expapi.JobStatus{}
```
Okay, good, this is clearing status.

erictune commented:
Why do several of the existing types have a `pkg/registry/$TYPE/registry.go`, but this PR does not?

erictune commented:
I'm back from vacation, and getting Job reviewed and merged in the next 6 days, so it can go into v1.1, is my top priority.

erictune commented:
Looks good aside from my and Clayton's comments.

soltysh commented:
@erictune @smarterclayton I've addressed your comments and updated tests, ptal.

soltysh commented on pkg/registry/job/strategy.go:
```diff
@@ -0,0 +1,110 @@
+	return fields.Set{
+		"metadata.name":              job.Name,
+		"status.successful":          strconv.Itoa(job.Status.Successful),
+		"status.conditions.complete": string(completeStatus),
```
@smarterclayton I'm not sure this is the right way to achieve what you wanted, to select complete jobs?

soltysh replied:
After discussing with @smarterclayton we've decided to drop this for now.

soltysh commented:
The failure is related to some check I'm missing from master, will rebase to pick up everything.

soltysh commented:
Rebased, tests should be ok now. For the record, previous commit with @smarterclayton comments was https://github.com/soltysh/kubernetes/commit/3ab923faad17da72f0c06fb5e9b4f59c78ba3c68

erictune commented:
LGTM

erictune commented:
Please squash commits.

soltysh commented:
Squashed.

karlkfi commented:
@soltysh  I see this got merged without docs changes. Are those gonna be part of #13254? Not sure where experimental APIs get documented, but we should figure it out.

erictune commented:
@karlkfi It seems we do not do automatic generation of static docs for experimental api group. We should. Filed #13883 for that.

karlkfi commented:
I was thinking more like a user-guide section for the new API/resource, but yes, those too.

soltysh commented:
@erictune thanks
@karlkfi I'll be adding samples as a separate PR to current ones. These are already quite big, I don't won't to clutter them more. I'll tag you in one.
